A complete overhaul of the cmake-scripts and the project structure

Febbe opened the pull request:
The current cmake scrips have a huge problem, they grew since the beginning of the project and both old cmake 2.x configurations as cmake 3.x configurations were used.    

A lot of good things have accumulated over time. However, some old / deprecated or wrong configurations have remained, which now cause problems.
For example, all sources are added manually to the targets instead of collecting them in a target and linking them against it.
This affects utilities that work on the targets.
Explicitly clang-tidy currently processes files twice, so fixes are applied 2 times.
Also, the CMake scripts contain hardcoded configurations which should be defined by the caller of CMake.
   
This PR addresses this and tidies up everything.
   
Another thing is, that there are 2 main targets: `util` and `xournalpp-core` but both require files from each other.
All files in `util`, depending on `xournalpp-core` were moved to the corresponding target.
Also, those targets are splitted strictly.

To make this separation more clear to others, the filesystem structure has been reordered too.
It looks like:
```text
xournalpp
    ┝ ...
    ┝ src
        ┝ CMakeLists.txt
        ┝ [some CMake generator files]
        ┝ core
            ┝ CMakeLists.txt 
            ┝ <core-modules>
        ┝ util
            ┝ CMakeLists.txt
            ┝ include
                ┖ util
                    ┖ <public-includes>
            ┝ [sources]
    ┝ CMakeLists.txt
```

Technius reviewed (changes requested):
Sorry, but I don't think this is ready for review yet.

* The commits need to be broken into pieces such that the build file refactor and the file renaming are separate commits. No one has time to review 963 files. In fact, it would be better if the file renaming is done in a separate PR (if we even want to rename files at all...).
* The reorganization will cause massive merge conflicts with existing PRs. We should make sure all PRs are either merged or rejected before we do any reorganization.
* There is no build time improvement based on CI build logs, and this is probably not the place to look for first for build time optimization. Most of the build time is spent invoking the compiler, not in configuring the build.
* Please explain what "cmake script issues" are being fixed.

Febbe commented:
> Sorry, but I don't think this is ready for review yet.
 
 Sorry, but your short overview is everything I need for the beginning.
 I wanted you to review this, to prevent me from investigating too much work into a wrong direction.
 
 > The commits need to be broken into pieces such that the build file refactor and the file renaming are separate commits.
 > The reorganization will cause massive merge conflicts with existing PRs. We should make sure all PRs are either merged or rejected before we do any reorganization.

They (mostly) are separated to make merges/rebases automatically possible. 
I am aware of that issue with the PR's. My target is to work on this, until it's clean enough. I want to continuously rebase it onto the current master. The merge should happen, as soon there are only small PR's.

> There is no build time improvement based on CI build logs, and this is probably not the place to look for first for build time optimization. Most of the build time is spent invoking the compiler, not in configuring the build.

Thought that there will be some, since parts have been built twice. 

> Please explain what "cmake script issues" are being fixed.

Will update the description.

LittleHuba commented on src/util/CMakeLists.txt:
```diff
@@ -4,9 +4,9 @@ file (GLOB_RECURSE util_SOURCES *.cpp *.h)
         )
+target_compile_features(util PUBLIC ${PROJECT_CXX_FEATURES})
+target_include_directories(util PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+add_library(xoj::util ALIAS util)
```
Why make this an alias if you could just use the actual name throughout the file?

LittleHuba commented on src/core/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+        PUBLIC xoj::defaults xoj::external_modules std::filesystem cxx17
+        )
+target_include_directories(xournalpp-core PUBLIC ${xournalpp_include_dirs})
+add_library(xoj::core ALIAS xournalpp-core)
```
Why make this an alias if you could just use the actual name throughout the whole file?

Febbe replied:
Doesn't have a higher meaning. But it moves our targets into a namespace and makes them read only. They are also visible everywhere, even in top level CMakeLists.txt.

LittleHuba commented on src/CMakeLists.txt:
```diff
@@ -1,106 +1,42 @@
+# xournalpp main program
+##############################
+
+add_executable(xournalpp core/Xournalpp.cpp)
```
If the file is in the core directory it should also be part of the core target. Either refactor this back out to the main directory or move this call to the cmake script within the directory and use [this property](https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html#prop_tgt:RUNTIME_OUTPUT_DIRECTORY) to build the binary in the correct directory.

We should also consider using `WIN32` and `MACOSX_BUNDLE` in this call to improve support within Windows and OSX.

Febbe replied:
moved it to a dir named `exe/`

LittleHuba commented on src/CMakeLists.txt:
```diff
@@ -1,106 +1,42 @@
+##############################
+# util
+##############################
+
```
Nitpick: blank line should be below not above...

LittleHuba commented on CMakeLists.txt:
```diff
@@ -59,268 +58,251 @@ include (GitRepo)
+if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+    # Fix linker flag handling for MacOS builds where frameworks are separated from their flag
+    string(REPLACE "-framework;" "-framework " xournalpp_external_deps "${xournalpp_external_deps}")
+endif ()]]
```
Are you sure this fix is no longer required?

Febbe replied:
Not sure, but I think thats the job of cmake.

LittleHuba replied:
I had to include this fix way back because CMake did not do this. We have to check the pipeline for this and do some test runs.

Febbe replied:
Nah I think, this is obsolete after my fix / it's our fault, when this is necessary.

LittleHuba replied:
Then please remove it completely.

LittleHuba commented on CMakeLists.txt:
```diff
@@ -59,268 +58,251 @@ include (GitRepo)
-    # Do not add system headers and external libraries to the additional include paths
-    # set(xournalpp_INCLUDE_DIRS ${xournalpp_INCLUDE_DIRS} "${INCLUDES}")
-endmacro (add_includes_ldflags LDFLAGS INCLUDES)
+endmacro(add_includes_ldflags LDFLAGS INCLUDES)
```
This macro should be obsolete now.

Febbe replied:
done

LittleHuba reviewed (commented):
Nice work! This was long overdue. I assume that you did not change anything more than include paths within headers and source files. I just checked the CMake files and provided some comments.

Febbe commented on src/core/control/Actions.h:
```diff
@@ -21,6 +21,7 @@
 #include "enums/ActionType.enum.h"
 
 
+
```
```suggestion
```

LittleHuba commented:
I'll do the review once the build issues on Windows and MacOS are fixed. No real use in checking everything if you need to change something afterwards.

Febbe commented:
The problem is, that I can't fix windows, since I don't get why it fails on the server.

LittleHuba commented:
I assume this is the real cause of the error: 

```
C:\msys64\mingw64\bin\xgettext.exe: error while opening "../src/core/gui/toolbarMenubar/FontButton.pp" for reading: No such file or directory
```

Everything after that is just a product of missing files after this step.

Febbe commented:
Jep, but what causes some generator to strip the c of .cpp?

LittleHuba commented:
That file is also missing the minimum CMake version...

Febbe commented:
> 
> 
> That file is also missing the minimum CMake version...

There is also a lot more to do, but I don't want to do so much work in one PR.

LittleHuba commented on po/CMakeLists.txt:
```diff
@@ -1,26 +1,30 @@
+          GLADEFILES
+          "${PROJECT_SOURCE_DIR}/ui/*.glade"
+          )
+  message(STATUS "SRCFILES: ${SRCFILES}")
+  message(STATUS "potfile: ${potfile}")
```
It fails before these lines. They won't give you more information. Better to paste the content of Gettext here and annotate it in place.

Febbe replied:
Do you mean in Gettext.cmake?
Now it works, but the gettext error is still there.

Febbe replied:
the variable contains `../src/core/gui/toolbarMenubar/FontButton.cpp` so this happens in the call to xgettext.

Febbe replied:
I think this is either a xgettext, cmake, or ninja bug. It also happens in all our pipeline runs.

LittleHuba replied:
Yeah the bug seems to be related to either the CMake version or the change to PROJECT_NAME. I think it is the latter. For this PR it is okay to leave it as it is, since it does not break the full pipeline. We should definitely follow up on this though.

Febbe replied:
There are 2 bugs, one, which is fixed by replacing PROJECT_PACKAGE with PROJECT_NAME.
And another one, which was there before (not introduced by this PR). The latter is mostly a dependency issue.

LittleHuba replied:
Then this is resolved for now. Could you create a ticket for the latter issue?

Febbe replied:
https://github.com/xournalpp/xournalpp/issues/3305

LittleHuba commented:
Seems like either [044527](https://github.com/xournalpp/xournalpp/pull/2570/commits/04452721d309016a59462ba9a6e8eef35cc8bee5) or [36df16](https://github.com/xournalpp/xournalpp/pull/2570/commits/36df16d3ae3f921cd358a5584ce02320a9efe434) fixed it.

rolandlo commented:
With commit a30b0e7 I get the following error message on MacOS Catalina while building:
```
admin@admins-MacBook-Pro build % cmake --build .
[  1%] Build translations.
[ 18%] Built target translations
[ 24%] Built target util
[ 97%] Built target xournalpp-core
[ 97%] Linking CXX executable ../xournalpp
clang: error: no such file or directory: 'AudioToolbox'
clang: error: no such file or directory: 'AudioUnit'
clang: error: no such file or directory: 'Carbon'
make[2]: *** [src/CMakeFiles/xournalpp.dir/build.make:130: xournalpp] Error 1
make[1]: *** [CMakeFiles/Makefile2:304: src/CMakeFiles/xournalpp.dir/all] Error 2
make: *** [Makefile:172: all] Error 2
```

LittleHuba commented:
@rolandlo can you revert tha last three commits and try once more?

Febbe commented:
> With commit [a30b0e7](https://github.com/xournalpp/xournalpp/commit/a30b0e72cd498342b76a67255df8ca5b4fc40ea2) I get the following error message on MacOS Catalina while building:

Have you executed cmake again? I've moved lots of files, So this may be an issue here, that cmake doesn't noticed.

LittleHuba commented:
> There is also a lot more to do, but I don't want to do so much work in one PR.
It is already very large

As long as you do not move any files around, this can be postponed to another PR. But moving files around is a blocker for other PRs. So you should do this in one PR if possible.
#3299 has to wait for this PR as well. So this is somewhat pressing...

LittleHuba commented on CMakeLists.txt:
```diff
@@ -1,18 +1,18 @@
+        DESCRIPTION "Xournal++ - Open source hand note-taking program"
+        HOMEPAGE_URL "https://xournalpp.github.io/"
+        LANGUAGES CXX C)
+# according to https://cmake.org/cmake/help/latest/command/project.html fix suffix
```
Do you want to introduce tweaks? This should be solved now as modifications to the version system will break the pipelines and all helper scripts.

Febbe replied:
What do you mean? This does not introduce tweaks, since I don't set a tweak to something else than 0?

LittleHuba replied:
For what is this comment then?

Febbe replied:
Ah, to add our ~dev postfix, the hole logic after project is for that purpose.

LittleHuba replied:
Then I don't get why that comment is there. Is it just a description for the following lines or is it actually a TODO? For me it reads like the latter.

LittleHuba commented on CMakeLists.txt:
```diff
@@ -59,270 +58,254 @@ include (GitRepo)
+	# Todo use platform independent stacktrace library
+	# Currently this works only for linux systems
+	##############################
+    find_package(Backtrace REQUIRED)
```
You should not load this as required within some if clause if we do not actually require it after all...

LittleHuba commented on CMakeLists.txt:
```diff
@@ -59,270 +58,254 @@ include (GitRepo)
+    target_include_directories(lua INTERFACE ${LUA_INCLUDE_DIR})
+    target_compile_definitions(lua INTERFACE ENABLE_PLUGINS)
+    add_library(Lua::lua ALIAS lua)
+    #set(ENABLE_PLUGINS "true")
```
Remove it if it is no longer required

rolandlo commented:
> > With commit [a30b0e7](https://github.com/xournalpp/xournalpp/commit/a30b0e72cd498342b76a67255df8ca5b4fc40ea2) I get the following error message on MacOS Catalina while building:
> 
> Have you executed cmake again? I've moved lots of files, So this may be an issue here, that cmake doesn't noticed.

I dropped the last 3 commits, removed all files from the build folder. After ```cmake ..``` and ```cmake --build .``` I got the same error messages again.

LittleHuba commented on CMakeLists.txt:
```diff
@@ -332,52 +315,55 @@ Configuration:
+set(CPACK_STRIP_FILES ON) # Debian packages must be stripped from debug messages, so lintian doesn't complain
 
 # .deb package options
-set (CPACK_DEBIAN_PACKAGE_VERSION "${CPACK_PROJECT_VERSION}")
-set (CPACK_DEBIAN_PACKAGE_HOMEPAGE ${PACKAGE_URL})
-set (CPACK_DEBIAN_PACKAGE_MAINTAINER "Andreas Butti <andreasbutti@gmail.com>")
```
These lines should not be removed. We should update the maintainer though as Andreas has paused his work on Xournal++ indefinitely for now.

Febbe replied:
I don't know why I romoved them, but I think those are set by the project variables.

Febbe replied:
`CPACK_DEBIAN_PACKAGE_VERSION` will be set by default.
`CPACK_DEBIAN_PACKAGE_HOMEPAGE` will be set by default.
`CPACK_DEBIAN_PACKAGE_MAINTAINER` is set to `CPACK_PACKAGE_CONTACT`, but that variable does not exist,
so I propose we chose a new contact and set `CPACK_PACKAGE_CONTACT` to it.

LittleHuba replied:
Okay then lets set the contact and we are good. It has to be in the previously used format to be supported by Debian.

LittleHuba replied:
Still missing the update to the maintainer

LittleHuba commented on src/config.h.in:
```diff
@@ -21,7 +21,7 @@ constexpr auto PROJECT_NAME = "@PROJECT_NAME@";
  * Name of package
  */
-constexpr auto PROJECT_PACKAGE = "@PROJECT_PACKAGE@";
+constexpr auto PROJECT_PACKAGE = "@PROJECT_NAME@";
```
This is not really great. Rename the target variable as well.

Febbe replied:
The target variable already exists...

Febbe replied:
I just removed it

LittleHuba commented on test/CMakeLists.txt:
```diff
@@ -1,58 +1,44 @@
 ## Additional config ##
```
CMake version requirement missing

LittleHuba commented:
> > > With commit [a30b0e7](https://github.com/xournalpp/xournalpp/commit/a30b0e72cd498342b76a67255df8ca5b4fc40ea2) I get the following error message on MacOS Catalina while building:
> > 
> > 
> > Have you executed cmake again? I've moved lots of files, So this may be an issue here, that cmake doesn't noticed.
> 
> I dropped the last 3 commits, removed all files from the build folder. After `cmake ..` and `cmake --build .` I got the same error messages again.

Have you cleared the CMakeCache?

rolandlo commented:
The CMakeCache is fully generated in the build folder, right? Since I executed `rm -r *` in the build folder, I think the CMakeCache was cleared.

LittleHuba commented:
> The CMakeCache is fully generated in the build folder, right? Since I executed `rm -r *` in the build folder, I think the CMakeCache was cleared.

No hidden files remain. Use `rm -r .*` instead or remove the full directory.

rolandlo commented:
Same error if I remove the build folder entirely. 
Something similar happened in this [godot issue](https://github.com/godotengine/godot/issues/36720)

LittleHuba commented:
> Same error if I remove the build folder entirely.
> Something similar happened in this [godot issue](https://github.com/godotengine/godot/issues/36720)

@Febbe  that was the fix I introduced and you removed

Febbe commented:
> 
> 
> > Same error if I remove the build folder entirely.
> > Something similar happened in this [godot issue](https://github.com/godotengine/godot/issues/36720)
> 
> @Febbe that was the fix I introduced and you removed

The variable does not even exist anymore, since we now work on targets, not library lists.
Also, your fix replaces a list with a space separated string.
I wounder, which library does not include his dependencies correctly.

Febbe commented:
For reference https://cmake.org/pipermail/cmake/2016-December/064711.html

Febbe commented:
@rolandlo which cmake version do you use on your mac?

rolandlo commented:
> @rolandlo which cmake version do you use on your mac?

version 3.17.2

Febbe commented:
Do you know, how I can make the pipeline fail as well?

Febbe commented:
> 
> 
> > @rolandlo which cmake version do you use on your mac?
> 
> version 3.17.2

This version is even too high to bump the minimal version, then 3.18 or 3.20 will be way too high, even it might be fixed there tbh.

Febbe commented:
@rolandlo can you retry and post the output of the `PkgConfig::ExternalModules INTERFACE_LINK_LIBRARIES` variable?

rolandlo commented:
I am getting lots of error messages now. 

```
CMake Error at CMakeLists.txt:143 (get_property):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "RULE_LAUNCH_COMPILE" is not allowed.
Call Stack (most recent call first):
  CMakeLists.txt:151 (print_target_properties)
```

etc. and at the end

```
CMake Error at CMakeLists.txt:143 (get_property):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "XCTEST" is not allowed.
Call Stack (most recent call first):
  CMakeLists.txt:151 (print_target_properties)


-- Could NOT find X11 (missing: X11_X11_INCLUDE_PATH X11_X11_LIB) 
help2man is missing; no man pages will be generated

Configuration:
        Compiler:                   /Library/Developer/CommandLineTools/usr/bin/g++
        CppUnit enabled:            OFF
        Filesystem library:         std::filesystem

-- Configuring incomplete, errors occurred!
See also "/Users/admin/dev/xournalpp/build/CMakeFiles/CMakeOutput.log".
```

rolandlo commented:
Now I am getting the following results:

```
-- PkgConfig::ExternalModules::INTERFACE_LINK_LIBRARIES: /Users/admin/gtk/inst/lib/libgtk-3.dylib;/Users/admin/gtk/inst/lib/libgdk-3.dylib;/Users/admin/gtk/inst/lib/libpangocairo-1.0.dylib;/Users/admin/gtk/inst/lib/libpango-1.0.dylib;/Users/admin/gtk/inst/lib/libharfbuzz.dylib;/Users/admin/gtk/inst/lib/libatk-1.0.dylib;/Users/admin/gtk/inst/lib/libcairo-gobject.dylib;/Users/admin/gtk/inst/lib/libpoppler-glib.dylib;/Users/admin/gtk/inst/lib/libgthread-2.0.dylib;/Users/admin/gtk/inst/lib/libxml2.dylib;/Users/admin/gtk/inst/lib/libzip.dylib;/Users/admin/gtk/inst/lib/libportaudiocpp.dylib;/Users/admin/gtk/inst/lib/libportaudio.dylib;/Users/admin/gtk/inst/lib/libsndfile.a;/Users/admin/gtk/inst/lib/librsvg-2.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libm.tbd;/Users/admin/gtk/inst/lib/libgio-2.0.dylib;/Users/admin/gtk/inst/lib/libgdk_pixbuf-2.0.dylib;/Users/admin/gtk/inst/lib/libgobject-2.0.dylib;/Users/admin/gtk/inst/lib/libglib-2.0.dylib;/Users/admin/gtk/inst/lib/libintl.dylib;/Users/admin/gtk/inst/lib/libcairo.dylib
-- PkgConfig::ExternalModules::INTERFACE_LINK_LIBRARIES fixed: /Users/admin/gtk/inst/lib/libgtk-3.dylib;/Users/admin/gtk/inst/lib/libgdk-3.dylib;/Users/admin/gtk/inst/lib/libpangocairo-1.0.dylib;/Users/admin/gtk/inst/lib/libpango-1.0.dylib;/Users/admin/gtk/inst/lib/libharfbuzz.dylib;/Users/admin/gtk/inst/lib/libatk-1.0.dylib;/Users/admin/gtk/inst/lib/libcairo-gobject.dylib;/Users/admin/gtk/inst/lib/libpoppler-glib.dylib;/Users/admin/gtk/inst/lib/libgthread-2.0.dylib;/Users/admin/gtk/inst/lib/libxml2.dylib;/Users/admin/gtk/inst/lib/libzip.dylib;/Users/admin/gtk/inst/lib/libportaudiocpp.dylib;/Users/admin/gtk/inst/lib/libportaudio.dylib;/Users/admin/gtk/inst/lib/libsndfile.a;/Users/admin/gtk/inst/lib/librsvg-2.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libm.tbd;/Users/admin/gtk/inst/lib/libgio-2.0.dylib;/Users/admin/gtk/inst/lib/libgdk_pixbuf-2.0.dylib;/Users/admin/gtk/inst/lib/libgobject-2.0.dylib;/Users/admin/gtk/inst/lib/libglib-2.0.dylib;/Users/admin/gtk/inst/lib/libintl.dylib;/Users/admin/gtk/inst/lib/libcairo.dylib
CMake Error at CMakeLists.txt:116 (set_target_properties):
  set_target_properties called with incorrect number of arguments.
```

Febbe commented:
Hopefully this will work for older cmake versions. If not, I would drop the support for those versions.

rolandlo commented:
> Hopefully this will work for older cmake versions. If not, I would drop the support for those versions.

I think we still need to support the `cmake` version of Ubuntu 18.04, which is `cmake 3.10`. In Ubuntu 20.04 it is `cmake 3.16`.

Febbe commented:
@rolandlo, can you retest this please?

Febbe commented:
@rolandlo bumbing cmake is not such an issue for us, because it is only a tool of the Toolchain. Nothing we require to link to. We could even bumb the compiler to clang-13 if we consequently crosscompile for clang-9. But the later would be a big overhead for us😅.

rolandlo commented:
Now during `cmake ..` I get:
```
-- PkgConfig::ExternalModules::INTERFACE_LINK_LIBRARIES: /Users/admin/gtk/inst/lib/libgtk-3.dylib;/Users/admin/gtk/inst/lib/libgdk-3.dylib;/Users/admin/gtk/init/lib/libpangocairo-1.0.dylib;/Users/admin/gtk/inst/lib/libpango-1.0.dylib;/Users/admin/gtk/inst/lib/libharfbuzz.dylib;/Users/admin/gtk/inst/lib/libatk-1.0.dylib;/Users/admin/gtk/inst/lib/libcairo-gobject.dylib;/Users/admin/gtk/inst/lib/libpoppler-glib.dylib;/Users/admin/gtk/inst/lib/libgthread-2.0.dylib;/Users/admin/gtk/inst/lib/libxml2.dylib;/Users/admin/gtk/inst/lib/libzip.dylib;/Users/admin/gtk/inst/lib/libportaudiocpp.dylib;/Users/admin/gtk/inst/lib/libportaudio.dylib;/Users/admin/gtk/inst/lib/libsndfile.a;/Users/admin/gtk/inst/lib/librsvg-2.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libm.tbd;/Users/admin/gtk/inst/lib/libgio-2.0.dylib;/Users/admin/gtk/inst/lib/libgdk_pixbuf-2.0.dylib;/Users/admin/gtk/inst/lib/libgobject-2.0.dylib;/Users/admin/gtk/inst/lib/libglib-2.0.dylib;/Users/admin/gtk/inst/lib/libintl.dylib;/Users/admin/gtk/inst/lib/libcairo.dylib
-- PkgConfig::ExternalModules::INTERFACE_LINK_LIBRARIES fixed: /Users/admin/gtk/inst/lib/libgtk-3.dylib;/Users/admin/gtk/inst/lib/libgdk-3.dylib;/Users/admin/gtk/inst/lib/libpangocairo-1.0.dylib;/Users/admin/gtk/inst/lib/libpango-1.0.dylib;/Users/admin/gtk/inst/lib/libharfbuzz.dylib;/Users/admin/gtk/inst/lib/libatk-1.0.dylib;/Users/admin/gtk/inst/lib/libcairo-gobject.dylib;/Users/admin/gtk/inst/lib/libpoppler-glib.dylib;/Users/admin/gtk/inst/lib/libgthread-2.0.dylib;/Users/admin/gtk/inst/lib/libxml2.dylib;/Users/admin/gtk/inst/lib/libzip.dylib;/Users/admin/gtk/inst/lib/libportaudiocpp.dylib;/Users/admin/gtk/inst/lib/libportaudio.dylib;/Users/admin/gtk/inst/lib/libsndfile.a;/Users/admin/gtk/inst/lib/librsvg-2.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/lib/libm.tbd;/Users/admin/gtk/inst/lib/libgio-2.0.dylib;/Users/admin/gtk/inst/lib/libgdk_pixbuf-2.0.dylib;/Users/admin/gtk/inst/lib/libgobject-2.0.dylib;/Users/admin/gtk/inst/lib/libglib-2.0.dylib;/Users/admin/gtk/inst/lib/libintl.dylib;/Users/admin/gtk/inst/lib/libcairo.dylib
-- Found ZLIB: /Users/admin/gtk/inst/lib/libz.dylib (found version "1.2.11") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Could NOT find X11 (missing: X11_X11_INCLUDE_PATH X11_X11_LIB) 
-- Found Lua: /usr/local/lib/liblua.a (found suitable version "5.3.6", minimum required is "5.3") 
Enable Xournal++ Plugins
-- Found Gettext: /Users/admin/gtk/inst/bin/xgettext (found version "0.20.2") 
help2man is missing; no man pages will be generated

Configuration:
        Compiler:                   /Library/Developer/CommandLineTools/usr/bin/g++
        CppUnit enabled:            OFF
        Filesystem library:         std::filesystem

-- Configuring done
-- Generating done
-- Build files have been written to: /Users/admin/dev/xournalpp/build
```

The linking after `cmake --build .` has arrived at 97% fails like before though.

Febbe commented:
`PkgConfig::ExternalModules::INTERFACE_LINK_LIBRARIES` is missing all the `-framework; X`'s. Maybe it is in another property.
But, I think, using cmake 3.18 on macOS wouldn't be a problem. Can you find out, what's in the other properties? Or do you agree with bumping the minimal required version for macOS?

LittleHuba commented:
I would say bumping CMake is the least effort and okay for MacOS. There is not a huge developer community running Macs and those people normally know how to install a newer version of CMake. All other users will never touch this subject.

rolandlo commented:
> I would say bumping CMake is the least effort and okay for MacOS. There is not a huge developer community running Macs and those people normally know how to install a newer version of CMake. All other users will never touch this subject.

What's the easiest way to upgrade cmake then? The one built via `jhbuild` inside the `Gtk` folder has version 3.17 and running `jhbuild buildone cmake` does not produce a newer version.

LittleHuba commented:
How about https://cmake.org/download/

Febbe commented:
@rolandlo you need to rerun the complete jhbuild upgrade. Buildone will only upgrade, if no other dependencies have to be upgraded.

rolandlo commented:
Ok, I am upgrading then.

rolandlo commented:
Fine, it compiles now on MacOS, after starting all over with `jhbuild`. I have now `cmake` 3.20.0 and `Gtk+` 3.24.30. Building Poppler via the `build-popper.sh` script didn't work any more (due to some invalid binding of a volatile rvalue https://gitlab.freedesktop.org/poppler/poppler/-/issues/1060). The script downloads version 0.72 of Poppler. I upgraded it to the latest version 21.08, which solved the problem. I also had some troubles with invalid `locale` (which I solved by exporting `LANG` and `LC_ALL` with value `en_US.UTF-8`) and with `jhbuild` not recognising the version of my Xcode command line tools. I solved that by hacking the `jhbuildrc` configuration file (returning the correct version 12.4 on my system manually). On the way I also had to install rust.

I noticed that now after building Xournal++ with 
```
cmake ..
cmake --build .
````
the app is not placed in `build/src` but in `build` directly. Is that intentional?

The `complete-build.sh` script worked as well.

Do I need to test on another OS than macOS Catalina?

LittleHuba commented:
Nice!

> the app is not placed in build/src but in build directly. Is that intentional?

Yes that is intentional.

Can you provide the updated build-poppler.sh so we can try it on the pipeline? Would be nice to update it there as well.

rolandlo commented:
> Can you provide the updated build-poppler.sh so we can try it on the pipeline? Would be nice to update it there as well.

The `build-poppler.sh` script is not run on the pipeline. It is only used to build the `gtk` folder that is used locally or fetched on the pipeline from the `xournalpp-pipeline-dependencies` repository https://github.com/xournalpp/xournalpp-pipeline-dependencies.
The only thing I modified in the script is the version number for Poppler (`21.08.0` instead of `0.72.0`). There are other dependencies with hard-coded version numbers (like the one from `openjpeg` 2.3.0, for which version 2.4.0 has been released https://github.com/uclouvain/openjpeg/releases; also `portaudio`, `libzip` and `cppunit` in some other scripts for MacOS), that might be better to update.

LittleHuba commented:
Ahh yeah now I remember. There is still the open PR for macports. Maybe we should switch to this setup in the pipeline as well. Because I have no Mac, there is no way I can do this. Would you be interested in trying your hand at a port?

Febbe commented:
Until macports also ports X11 this is not a good idea. brew also has this issue.

I would also use brew on Mac for our normal pipeline runs, and only use jhbuild/macports for the releases.

LittleHuba commented:
You are both more than welcome to work on this. I think it is best to tackle this right after this PR. That is also why I brought I up initially. With the updated CMake scripts, the improved pipelines and the helper scripts we should be able to streamline the release process and testing of PRs for all OSs with up to date dependencies.

rolandlo commented:
> Until macports also ports X11 this is not a good idea. brew also has this issue.
> 
Can you explain why this is a problem?
jhbuild may have the advantage, that one can build pretty recent package versions with it. I don't think this is the case with macports. For example via jhbuild I get Gtk+ 3.24.30 (same for brew). The macports version is only 3.24.23, https://ports.macports.org/port/gtk3/. On the other hand jhbuild is also a pain to update, since there are new issues every time.

With brew I don't know how to build a standalone app-bundle. Even with macports it seems problematic, see https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues/8

Febbe commented:
> Can you explain why this is a problem?

Well, it's not really a problem. Maybe it works well. But it brings another huge dependency and another indirection layer with it.
It would be better, to just use the native window manager of the system we are on. And gtk supports the macOS windowing system.

> With brew I don't know how to build a standalone app-bundle. Even with macports it seems problematic, see https://gitlab.gnome.org/GNOME/gtk-mac-bundler/-/issues/8

Jep, this is indeed a problem for both. But I think that's not a problem, when using only one of the 3: macports, jhbuild or brew.

rolandlo commented:
Are you sure it is not a problem to create a standalone app-bundle when building with brew as long as it is not mixed with `jhbuild` and `macports`? The author of the `gtk-mac-bundler` says:
> Although I have heard of people managing to bundle applications built with MacPorts it's apparently non-trivial; I've never heard of anyone managing it with Homebrew.

LittleHuba commented:
See #2390 for more details. The checks there did give the impression, that the application created is portable and does not require a macports installation.

LittleHuba commented:
To get back on topic, I think this PR is mostly ready to merge. We should resolve the merge conflict and maybe outsource the version information to a separate file as proposed in #3310

Febbe commented:
> To get back on topic, I think this PR is mostly ready to merge. We should resolve the merge conflict and maybe outsource the version information to a separate file as proposed in #3310

I would do that in a separate PR. Also, we should wait until 1.1.1 is released.

rolandlo commented:
I've tested this PR on MS Windows and haven't found any problems.

Febbe commented:
@xournalpp/core I rebased it.
Does anyone know why the pipeline is skipped?
Also, we should test all installers still working.

LittleHuba commented:
> @xournalpp/core I rebased it.
> Does anyone know why the pipeline is skipped?
> Also, we should test all installers still working.
> 

Check the logs on Azure. Either there are too many parallel jobs (just wait) or the pipeline scripts have a bug and pipelines can't start.

Febbe commented:
/azp run create-installers

Febbe commented:
Will merge this in 24h if no objections occur, and the pipelines succeed.

LittleHuba commented on CMakeLists.txt:
```diff
@@ -65,198 +60,180 @@ include (GitRepo)
-	find_package(Backtrace REQUIRED)
-	add_includes_ldflags("${Backtrace_LIBRARIES}" "${Backtrace_INCLUDE_DIRS}")
+    # libexec
+    # Todo does not belong here
```
Could you either fix this or describe what needs to be done? This way following up on your work is hard.

LittleHuba commented:
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY4ODEvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY4ODEvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY4ODEvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 18.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY4ODEvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0xOC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMTY4ODEvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).

LittleHuba commented on CMakeLists.txt:
```diff
@@ -267,70 +244,65 @@ add_subdirectory (man)
+    # Install icons
+    install(FILES ui/pixmaps/com.github.xournalpp.xournalpp.svg
+            DESTINATION share/icons/hicolor/scalable/apps)
+    # Symlink are not easy to use with CMake, therefore simple install a copy...
```
You can drop that comment as it holds no benefit to the reader.

Febbe commented:
@LittleHuba 
> Still missing the update to the maintainer

This variable defaults to `CPACK_PACKAGE_CONTACT` and is set to `"${PROJECT_CONTACT}"`, which is set above:
`set(PROJECT_CONTACT "core.xournalpp@github.com")`

Technius commented:
Hopefully this won't cause too many merge conflicts in #2379 and #2809.

Febbe commented:
@Technius I don't think so, git is pretty smart already. But at least #2379 can be merged before; as I know, it's finished.
@idotobi what's your status there?

bhennion commented:
I think we should wait for release-1.1 to be merged before merging this. This will create a lot of conflicts, and whoever handles the merging of release-1.1 will suffer.

Also, for the other PRs, @Febbe, could you document what files where moved where? It would help.

Febbe commented:
> I think we should wait for release-1.1 to be merged before merging this. This will create a lot of conflicts, and whoever handles the merging of release-1.1 will suffer.

Either @LittleHuba or me should do this I think. So this depends on, if @LittleHuba want to resolve the conflicts.
The conflicts must be resolved either way.

Also, git can differentiate the renaming of files. Therefore, we will have only normal merge conflicts.
A problem will arise for new or deleted files. But those are really easy to handle (just re add or remove them again)

@bhennion: 
A rule of thumb: I 
 - moved src/win32/* into src/exe/win32/*
 - moved headers of /src/util/* into /src/util/include/util/*
 - moved the the rest of src/* (excluding util and win32) into src/core/*
 
Complicated conflicts will arise for all CMakeLists.txt.
Easy conflicts for all src and header files (only include paths changed).
Very easy conflicts for deleted/removed files.

Febbe commented:
I honestly don't want to wait any longer, since I had to rebase/merge literally all merged PRs since last year.

bhennion commented:
> > I think we should wait for release-1.1 to be merged before merging this. This will create a lot of conflicts, and whoever handles the merging of release-1.1 will suffer.
> 
> Either @LittleHuba or me should do this I think. So this depends on, if @LittleHuba want to resolve the conflicts. The conflicts must be resolved either way.
> 
> Also, git can differentiate the renaming of files. Therefore, we will have only normal merge conflicts. A problem will arise for new or deleted files. But those are really easy to handle (just re add or remove them again)
> 
> @bhennion: A rule of thumb: I
> 
> * moved src/win32/* into src/exe/win32/*
> * moved headers of /src/util/* into /src/util/include/util/*
> * moved the the rest of src/* (excluding util and win32) into src/core/*
> 
> Complicated conflicts will arise for all CMakeLists.txt. Easy conflicts for all src and header files (only include paths changed). Very easy conflicts for deleted/removed files.

Alright, LGTM then!
